### PR TITLE
Change S(k) normalization to 1/N

### LINKED ISSUE #789: Diffraction normalization should be S(k=0) = N
We should change the normalization criteria of the DiffractionPattern such that `S(k=0) = N`. Currently it is `S(k=0) = 1`. The typical formulation in physics (see [Wikipedia](https://en.wikipedia.org/wiki/Structure_factor)) gives a result of `N` but we changed it to `1` so it's easier to plot. However, we can just adapt the plotting code to re-scale as needed, and keep the normal physics convention.

_Originally posted by @bdice in https://github.com/glotzerlab/freud/pull/777#issuecomment-853358425_

### PR DESCRIPTION
**Pull request by andkerr**
## Description
This pull request addresses issue #789, normalizing the structure factor computed with `freud.diffraction.DiffractionPattern` such that S(k=0) == N, for a system of N particles.

The initial commit only changes the normalization in diffraction.pyx. I used some basic tests to verify this change, but have yet to update the actual test files, the plotting function, or any other diffraction code.

This leaves a few to-do's before this pull request will be ready for merging:

- [x] Update the diffraction plot color scale
- [x] Adjust any relevant diffraction test assert statements to account for the extra factor of N
- [x] Update documentation and example notebooks

## Motivation and Context
This update brings freud's diffraction code in line with the usual physics convention that S(k=0) == N for a system of N particles.
Resolves: #789 

## How Has This Been Tested?
The following code tests a a simple cubic system of 4096 particles, viewed down the z-axis (grid_size = output_size = 512, zoom = 4). A screenshot of the resulting plot is included below.

normalization_test

## Screenshots

new_plot

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [ ] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [ ] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

### CODE REVIEW

**Comment by bdice** [comment]
@andkerr It looks like there are changes to git submodules which may not be related to this PR. Please double-check that and ping me whenever you're ready for a review.

**Comment by andkerr** [comment]
@bdice I just removed the changes to freud's submodules.

**Review by tcmoore3 — APPROVED** [review]

**Review by DomFijan — CHANGES_REQUESTED** [review]
Tests are failing. In particular:
```
  =========================== short test summary info ============================
  FAILED test_diffraction_DiffractionPattern.py::TestDiffractionPattern::test_center_unordered
  FAILED test_diffraction_DiffractionPattern.py::TestDiffractionPattern::test_center_ordered
  ============ 2 failed, 360 passed, 4 skipped in 1600.84s (0:26:40) =============
```
You should modify the two tests above to reflect the changes in the normalization.
Other then that seems fine.

**Review by bdice — CHANGES_REQUESTED** [review]
> This leaves a few to-do's before this pull request will be ready for merging:
>
> - [x] Update the diffraction plot color scale
> - [x] Adjust any relevant diffraction test assert statements to account for the extra factor of N
> - [x] Update documentation and example notebooks

There are some items from the PR description (fixing normalization for plotting functions and updating the documentation) that should be completed before we merge this PR.

Please post a screenshot of a diffraction pattern with the plotting code once it is updated.

**Comment by andkerr** [comment]
The latest commit adjusts freud/diffraction.pyx and freud/plot.py to make diffraction plots account for normalization by N. The color bar is changed from a logarithmic to a linear scale, and ticks are labelled with integer fractions of N, the number of points in the system. I added N to the plot title to reflect the the relationship between the value of the diffraction peaks and the number of points in the system (particularly at k = 0). See the attached screenshot, which is an FCC system (grid_size = output_size = 512, zoom = 4).

Screen Shot 2021-09-23 at 6 03 18 PM

**Comment by andkerr** [comment]
Latest commits update tests to make them pass, and bring documentation in line with the new convention S(k=0) = N.  Plots look more like what they do in the prior version, but the plotting code seems buggy and produces different results when dp.compute is run in example notebooks.

**Comment by andkerr** [comment]
The latest commits:
- revert the diffraction plot colormap to a logarithmic scale
- adjust vmin and vmax plotting parameters to values that produce cleaner plots
- remove the number of points from plot titles
- set color bars to use only two labelled ticks (0 and N)
- add the member variable `N_points` to the `DiffractionPattern` class as a computed property
- update docs and testing

The re-normalized plots better approximate the versions in the master branch diffraction example notebook:

### fcc system, N = 4000
Screen Shot 2021-11-04 at 2 35 34 PM

### the same system viewed from a different angle
Screen Shot 2021-11-04 at 2 35 51 PM

### random system, N = 10000
Screen Shot 2021-11-04 at 2 36 01 PM

@DomFijan The failing tests that we discussed last week now pass when I run them locally

**Comment by andkerr** [comment]
Questions that I still have about this PR:
- Are the plots accurate enough, or at least as good as what the latest Freud release generates?
- With only two values on the colorbar axis, are intermediate tick marks necessary?

File: freud/diffraction.pyx
Comment (bdice) [inline freud/diffraction.pyx]:
```suggestion
        # Transform the image (scale, shear, zoom) and normalize S(k) by N_points
```

File: freud/diffraction.pyx
Comment (bdice) [inline freud/diffraction.pyx]: This is a nice solution. Just remove the spaces between `` :code:`None` ``
```suggestion
            vmin (float):
                Minimum of the color scale. Uses :code:`4e-6 * N_points` if
                not provided or :code:`None` (Default value = :code:`None`).
            vmax (float):
                Maximum of the color scale. Uses :code:`0.7 * N_points` if
                not provided or :code:`None` (Default value = :code:`None`).
```

File: freud/diffraction.pyx
Comment (bdice) [inline freud/diffraction.pyx]: Use the public API property.
```suggestion
        if vmin is None:
            vmin = 4e-6 * self.N_points

        if vmax is None:
            vmax = 0.7 * self.N_points
```

File: freud/diffraction.pyx
Comment (bdice) [inline freud/diffraction.pyx]:
```suggestion
            vmin (float):
                Minimum of the color scale. Uses :code:`4e-6 * N_points` if
                not provided or :code:`None` (Default value = :code:`None`).
            vmax (float):
                Maximum of the color scale. Uses :code:`0.7 * N_points` if
                not provided or :code:`None` (Default value = :code:`None`).
```

File: freud/diffraction.pyx
Comment (bdice) [inline freud/diffraction.pyx]:
```suggestion
        if vmin is None:
            vmin = 4e-6 * self.N_points

        if vmax is None:
            vmax = 0.7 * self.N_points
```

File: freud/diffraction.pyx
Comment (bdice) [inline freud/diffraction.pyx]:
```suggestion
            self.diffraction, self.k_values, self.N_points,
```

File: freud/plot.py
Comment (bdice) [inline freud/plot.py]:
```suggestion
        vmin (float):
            Minimum of the color scale. Uses :code:`4e-6 * N_points` if
            not provided or :code:`None` (Default value = :code:`None`).
        vmax (float):
            Maximum of the color scale. Uses :code:`0.7 * N_points` if
            not provided or :code:`None` (Default value = :code:`None`).
```

File: freud/plot.py
Comment (bdice) [inline freud/plot.py]: I don't think we want to use the pyplot (`plt`) interface here - its behavior is different than the object-oriented API, I think it adds Axes objects or something weird. You might be able to revert to the previous code but adapt this example to set the ticks? https://matplotlib.org/stable/gallery/ticks_and_spines/colorbar_tick_labelling_demo.html

See #533, #318 for some relevant past work on plotting functionality.

File: tests/test_diffraction_DiffractionPattern.py
Comment (bdice) [inline tests/test_diffraction_DiffractionPattern.py]: Similarly below.
```suggestion
                    # The value at k=0 should be N_points, per convention.
```

File: tests/test_diffraction_DiffractionPattern.py
Comment (bdice) [inline tests/test_diffraction_DiffractionPattern.py]:
```suggestion
                    # The value at k=0 should be N_points, per convention.
```

**Review by bdice — CHANGES_REQUESTED** [review]
@andkerr I'm suggesting some small tweaks but otherwise this is good to go. Only one change will require testing manually again, relating to the use of `plt.colorbar`.

**Comment by andkerr** [comment]
Incorporated Bradley's suggested tweaks, and added a test to verify that the N_points property is computed correctly per Dom's suggestion.

File: freud/plot.py
Comment (bdice) [inline freud/plot.py]: The bottom bound isn't zero because this is a log scale. Also I agreed with @gaofy95's [comment](https://github.com/glotzerlab/freud-examples/pull/38#pullrequestreview-799091565) that it would be nice to have more than two tick values on the colorbar scale. I am going to push some changes to this section and then merge this PR.

Comment (bdice) [reply freud/plot.py]: Also the top bound is `0.7 * N_points`, not `N_points`. I am just going to revert this change to the colorbar and let it be a log scale like it was before. If we want to change plotting behavior further, let's do that in a separate PR, since everything else here is coherent/consistent for changing the normalization.

**Review by bdice — APPROVED** [review]
Great work! I am going to merge this once I check over CI and verify that the documentation renders well.

**Pull request merged**
